### PR TITLE
Unsupervised Classification of Solar Wind Source Regions

### PR DESCRIPTION
**Pull request by tbloch1**
This pull request is my contribtion to the HelioML book, covering my work on unsupervised machine learning for solar wind classification.

### CODE REVIEW

**Comment by mbobra** [comment]
@tbloch1 Thank you for this pull request! We really appreciate it 🙏

@jmason86 and I will take a look at this after the winter break. Happy holidays!

**Comment by jmason86** [comment]
Sorry for the really long time it took before review! Really abnormal for us. 

Review:
* Introduction looks perfect to me! Starts broad and understandable to anyone and drills down to the point in a clear and quick flow. 
* Table of contents update looks good. Follows the format, appears to be complete. 
* Plots look amazing! I use a dark theme on GitHub and they render well there as well as with a white background (which is how it displays in the book). This speaks to good color choice. Labels are all large, readable, and clear. 🙌 all hail `seaborn` 😃 And the purpose of the plots is well tied to the narrative text. 
* Narrative text is extensive and clear throughout. One particular highlight is the description of hyperparameters. Nice for the reader not to need to dig into documentation to get that information. 
* Code is clean and clear. Variable names appropriate to scope. 
--
∴ A+ on the high level stuff. I can't find anything to improve. 

Minor stuff: 

- [ ] Typo in intro to unsupervised learning: "Or, more straighforwardly, to accurately identify the all of the" -- extra "the" in there
- [ ] Typo in same paragraph: "we them analyse the clusters" -- "them" -> "then"
- [ ] Maybe unfixable, but in the "An Intuitive Classification Scheme" section, "our data in log$_{10}$-space for ease" renders weird, with a big space between the g and the 10. Could just be my system though, and may not render like that in the book. 
  * Same deal with the degree symbol when it's used (e.g., 180$^{\circ}$) and exponents. Guessing it's anytime that $$ is used.
  * I don't see anything wrong with that syntax so my best guess is that it is a GitHub rendering problem that shouldn't occur in the book
- [ ] Under the UMAP stuff, "but through various trials of different" appears to be a broken sentence. 
- [ ] In "Subsample Ensemble Results" section, there's a missing parenthesis here: "(coronal holes and the streamer-belt, respectively."
- [ ] Later in that section, this sentence looks like it's missing a "than": "This distribution is more Gaussian the full cluster 3 distribution."
- [ ] Could you promote your bolded sections to full headings (if they aren't already)? If you check out [any random chapter in the book](https://helioml.org/08/notebook.html), you'll see that headings become a high level guide on the right hand side, which is pretty handy. 
- [ ] If it's not hard to dig up, do you have a rough idea of how long it takes to execute this whole notebook? On a typical laptop/desktop or does it require a computational cluster? If you could provide a really brief summary in this regard either at the beginning or the end of the notebook, that would be appreciated. 

Just a question that occurred to me and maybe I missed the answer. Could you invert the process UMAP > 2D-Cluster to 6D-Cluster > UMAP? Would you expect the same basic end result? Not suggesting this go in the chapter, I'm just curious. 

As someone particularly interested in CMEs, I was especially intrigued by your finding that cluster 3 might be a novel ICME classifier. Really interesting analysis leading to it as well. 

Overall: Wow! I can't wait to get this available in the book. Super interesting and exceptionally well described. Thank you for the submission!

**Comment by tbloch1** [comment]
Hi James! 

Not to worry, the start of the year is usually a really busy time for everyone.

Thanks for such positive feedback, I appreciate it! I think I've managed to address all-but-one of you points now (the rendering of the $$ stuff should work now too, I think). I'm a bit confused about the sectioning though. Currently I use #, ##, ### for various levels of headings with a space between that and the heading text. As far as I could see online, this is the usual way to do it, but I must be missing something if it's not coming across on your end?

UMAP question:
When I initially started my PhD and did this work, there wasn't a way to map backwards since the mapping function was many ↦1. Recently though, that functionality was added in (not sure how it works), opening up the possibility. I haven't had a chance to test it out yet. My intuition would be that you would lose some of the information but retain some structure proportional to the structure you kept in the low-D representation. More interestingly though is that you can now sample a really high-D parameter space much more efficiently by using the low-D space that UMAP outputs (useful if you wanted to see how a model would train using data sampled from the whole low-D UMAP domain, rather than just sampling where you have data). I think it could be amazingly useful, but I'm still trying to wrap my head around exactly what it would do and how it would work.

ICMEs:
I hadn't really realised that people would be interested in the ICME aspect till I showed my supervisor the results that are in this chapter (they're an extension to the work in my paper). Since then I've thought that it would be cool to try to develop that a bit further, but my PhD and time is now commited elsewhere.

If you're particularly interested maybe we could have a chat about it at some point? I'd be keen to try to extend it, but really only have time to talk about ideas. But it would be cool to pass the torch on to someone else and see UMAP carve out more a place for itself in our field!

Anyway, let me know if anything else needs changing in the chapter, I think I have commited the extra stuff for you to check out.

Thanks again,
Téo

**Comment by jmason86** [comment]
Re: sectioning. Actually it does look right even in the GitHub render now. Hmpf. Must've just been a brainfart on my part. Thanks! 

I looked over the changes and they look great. 

The ICME stuff does seem like a good place for potential future collaboration. I know some other people who might be interested as well. I'm similarly overcommitted on things in the near term, but I'm going to keep this in mind and maybe we'll find a good opportunity to do something with it.

**Review by jmason86 — APPROVED** [review]
Excellent new chapter. I gave it a detailed review and all requested changes have been made.

**Comment by tbloch1** [comment]
That sounds great, I look forward to it!

Thanks for approving the chapter. Any idea when it will end up in the book (to share with other people etc.)?

**Comment by jmason86** [comment]
I expect it'll be in the book later today 🙂

**Comment by mbobra** [comment]
@tbloch1 Thanks for the submission and for all these edits, we really appreciate it! 
@jmason86 Thanks for the thorough review! I'm happy for this to be merged whenever you're ready.

**Review by mbobra — APPROVED** [review]

**Comment by mbobra** [comment]
@tbloch1 You chapter is live on [helioml.org](https://helioml.org/10/Unsupervised_Classification_of_Solar_Wind_Source_Regions.html)!

**Comment by tbloch1** [comment]
Thank you both again for all this and especially the quick turn-around time, very refreshing to have things all go so quickly!
